### DISCUSSION TITLE
Virtual rendering implementation behaves differently on Firefox, causing worse performance.

### DISCUSSION THREAD

**Discussion by emilio** [community/community#77953] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Bug

### Body

Over in [Mozilla bug 1868316](https://bugzil.la/1868316), we got a report of some kinds of files being much slower in Firefox compared to other browsers.

https://github.com/MaoJianwei/eAIP-2015-Nr.13/blob/master/Download_eAIP_2015_Nr.13.py is an extreme example of this.

While investigating, I realized that the DOM that GitHub ends up with is substantially (530x) larger on Firefox than on Chromium (829448 vs. 1556). That was done via `document.querySelector("*").length`, after also verifying there wasn't shadow DOM involved.

It seems there's something about the virtualization of the code view that doesn't work in Firefox, or at least works differently to Chromium? If it was due to something misbehaving in Firefox, I'd be happy to help diagnose and address it, but tentatively filing here.

Thank you!

**Comment by Cbodfield** [comment, marked as answer]
Hey @emilio!

Since the associated Mozilla bug is closed I am no longer able to comment on it. I just wanted to follow up to let you know that we've updated our code rendering to accommodate the `user-find` prototype you implemented. We are now doing `user-find` support detection and changing our rendering strategy to match non-Firefox browsers. 🎉

Additionally, prior to the `user-find` improvement, we did ship a few changes that should have helped performance a bit for standard Firefox users.

I appreciate you pushing the `user-find` efforts forward with the CSSWG! If you find any issues with our `user-find` implementation or have anything else we can help, please let us know.

**Reply by emilio** [reply]
That's awesome to hear, thank you!

**Reply by Cbodfield** [reply]
One more follow up -

As discussed, I've removed the `user-find` prototype since it didn't move forward with the CSSWG. I've also added usage of the `inert` attribute to hide code lines from browser find to go along with the Firefox-side behavior `inert` change.

Starting with Firefox 124, Firefox will have the same code rendering strategy as other browsers. 🥳 

This behavior can be tested / verified right now on Firefox Beta/Developer and Firefox Nightly.

Thanks again for all your help @emilio! It was great collaborating with you on this. ❤️

**Reply by emilio** [reply]
Same! Thanks for all your help with this!

**Reply by emilio** [reply]
For reference, Chromium folks filed https://issues.chromium.org/issues/325372936 for this, and I just filed https://bugs.webkit.org/show_bug.cgi?id=269909 for WebKit.

When all browsers implement this, you could simplify your rendering even further by avoiding all the pseudo-element shenanigans.

**Reply by AdamShwert** [reply]
Hey Emilio, thanks again for all of your help on this! In your experience, what is the approximate timeline for chromium/webkit to fix issues like this?

**Reply by emilio** [reply]
Hard to know, really depends on the complexity of the fix and the relative priority. My guess is that this shouldn't be a hard fix.

**Reply by AdamShwert** [reply]
Hopefully they give it high priority!
